divviup/src/main/java/org/divviup/android/Client.java

```diff
@@ -123,46 +162,43 @@ public void sendMeasurement(M measurement) throws IOException {
 
         String path = "tasks/" + this.taskId.encodeToString() + "/reports";
         URL url = leaderEndpoint.resolve(path).toURL();
-        HttpURLConnection connection = (HttpURLConnection)url.openConnection();
-        connection.setRequestMethod("PUT");
-        connection.setRequestProperty("User-Agent", getUserAgent());
-        connection.setRequestProperty("Content-Type", REPORT_CONTENT_TYPE);
-        connection.setDoOutput(true);
-        connection.setFixedLengthStreamingMode(report.length);
-        connection.connect();
-        OutputStream out = connection.getOutputStream();
-        IOUtils.write(report, out);
-        out.close();
-        int code = connection.getResponseCode();
-        if (code >= 400) {
-            throw new IOException(
-                    "aggregator returned HTTP response code " + code + " when uploading report"
-            );
+        RequestBody body = RequestBody.create(report, REPORT_CONTENT_TYPE);
+        Request request = new Request.Builder().url(url).put(body).build();
+        try (Response response = client.newCall(request).execute()) {
+            int code = response.code();
+            if (code >= 400) {
+                throw new IOException(
+                        "aggregator returned HTTP response code " + code + " when uploading report"
+                );
+            }
         }
     }
 
     private HpkeConfigList fetchHPKEConfigList(URI aggregatorEndpoint, TaskId taskId) throws IOException {
         String path = "hpke_config?task_id=" + taskId.encodeToString();
         URL url = aggregatorEndpoint.resolve(path).toURL();
-        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
-        connection.setRequestProperty("User-Agent", getUserAgent());
-        connection.setUseCaches(true);
-        connection.setDoInput(true);
-        connection.connect();
-        int code = connection.getResponseCode();
-        if (code >= 400) {
-            throw new IOException(
-                    "aggregator returned HTTP response code " + code + " when fetching HPKE configs"
-            );
-        }
-        String contentType = connection.getContentType();
-        if (!contentType.equals(HPKE_CONFIG_LIST_CONTENT_TYPE)) {
-            throw new IOException("wrong content type for HPKE configs: " + contentType);
+        Request request = new Request.Builder().url(url).build();
+        try (Response response = client.newCall(request).execute()) {
+            int code = response.code();
+            if (code >= 400) {
+                throw new IOException(
+                        "aggregator returned HTTP response code " + code + " when fetching HPKE configs"
+                );
```

Comment:
Throwing `IOException` in this function feels a bit strange to me. I interpret IOException as something retryable. (This isn't explicitly stated [in the docs](https://docs.oracle.com/javase/8/docs/api/java/io/IOException.html), but I would usually consider a transient I/O error as something that can be retried).

I think the cases in this function are all non-retryable and may benefit from throwing a distinct exception type.

I don't feel strongly about this, since prior code was using IOException anyways. It's just something to consider for future improvement. I'm also not familiar with what is standard practice in Java around exception handling.

Reply:
I don't think IOException necessarily indicates an error is retryable, rather it's an umbrella superclass for a large variety of exceptions. Back when I was using HttpUrlConnection, I threw an IOException with a custom message here because calling getInputStream() on a response with a non-200 status would itself throw an IOException. It looks like OkHttp has a strong preference for IOException as well: they expect interceptors/middleware to throw IOException for normal HTTP-level errors, but treat other thrown exceptions as a crash in the interceptor.